File: data-structures/non-linear/trees/binary-tree/iterative/BinaryTree.java
import java.util.*;

class Node {
  int data;
  Node left, right;
  Node(int data) {
    this.data = data;
    this.left = this.right = null;
  }
}

public class BinaryTree {
  public static void preOrder(Node root) {
    if (root == null)
      return;
    Stack<Node> s = new Stack<>();
    s.push(root);
    while (!s.isEmpty()) {
      Node n = s.pop();
      System.out.println(n.data + " ");
      if (n.right != null)
        s.push(n.right);
      if (n.left != null)
        s.push(n.left);
    }
  }

  public static void inOrder(Node root) {
    Stack<Node> s = new Stack<>();
    Node current = root;
    while (current != null || !s.isEmpty()) {
      while (current != null) {
        s.push(current);
        current = current.left;
      }
      current = s.pop();
      System.out.println(current.data + " ");
      current = current.right;
    }
  }

  public static void postOrder(Node root) {
    Stack<Node> s = new Stack<>();
    Node prev = null;
    Node current = root;
    while (current != null || !s.isEmpty()) {
      if (current != null) {
        s.push(current);
        current = current.left;
      } else {
        Node top = s.peek();
        if (top.right != null && prev != top.right) {
          current = top.right;
        } else {
          System.out.println(top.data + " ");
          prev = s.pop();
        }
      }
    }
  }

  public static void levelOrder(Node root) {
    if (root == null)
      return;
    Queue<Node> q = new LinkedList<>();
    q.offer(root);
    while (!q.isEmpty()) {
      Node node = q.poll();
      System.out.println(node.data + " ");
      if (node.left != null)
        q.offer(node.left);
      if (node.right != null)
        q.offer(node.right);
    }
  }
}

File: data-structures/non-linear/trees/binary-tree/iterative/BinaryTree.py
from collections import deque


class Node:
    def __init__(self, value):
        self.data = value
        self.left = None
        self.right = None


def pre_order(root):
    if not root:
        return

    stack = [root]
    while stack:
        node = stack.pop()
        print(node.data, end=" ")

        if node.right:
            stack.append(node.right)
        if node.left:
            stack.append(node.left)


def in_order(root):
    stack = []
    current = root

    while stack or current:
        if current:
            stack.append(current)
            current = current.left
        else:
            current = stack.pop()
            print(current.data, end=" ")
            current = current.right


def post_order(root):
    stack = []
    last_visited = None
    current = root

    while stack or current:
        if current:
            stack.append(current)
            current = current.left
        else:
            top = stack[-1]
            if top.right and last_visited != top.right:
                current = top.right
            else:
                print(top.data, end=" ")
                last_visited = stack.pop()


def level_order(root):
    if not root:
        return

    queue = deque([root])

    while queue:
        node = queue.popleft()
        print(node.data, end=" ")

        if node.left:
            queue.append(node.left)
        if node.right:
            queue.append(node.right)
